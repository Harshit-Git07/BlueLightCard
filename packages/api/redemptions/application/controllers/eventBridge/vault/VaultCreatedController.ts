import { z } from 'zod';

import { NON_NEGATIVE_INT } from '@blc-mono/core/schemas/common';
import { eventSchema } from '@blc-mono/core/schemas/event';
import { OPTIONAL_URL_SCHEMA } from '@blc-mono/core/schemas/utility';
import { Result } from '@blc-mono/core/types/result';
import { ILogger, Logger } from '@blc-mono/core/utils/logger/logger';
import { IVaultService, VaultService } from '@blc-mono/redemptions/application/services/vault/VaultService';
import { RedemptionsDatasyncEvents } from '@blc-mono/redemptions/infrastructure/eventBridge/events/datasync';

import { EventBridgeController, UnknownEventBridgeEvent } from '../EventBridgeController';

const VaultCreatedEventDetailSchema = z.object({
  adminEmail: z.string().email().nullable().optional(),
  alertBelow: z.number(),
  brand: z.string(),
  companyId: z.coerce.string(),
  companyName: z.string(),
  eeCampaignId: NON_NEGATIVE_INT.nullable().optional(),
  link: OPTIONAL_URL_SCHEMA,
  linkId: NON_NEGATIVE_INT.nullable().optional(),
  managerId: NON_NEGATIVE_INT.nullable().optional(),
  maxPerUser: NON_NEGATIVE_INT,
  offerId: z.coerce.string(),
  showQR: z.boolean(),
  ucCampaignId: NON_NEGATIVE_INT.nullable().optional(),
  vaultStatus: z.boolean(),
});
const VaultCreatedEventSchema = eventSchema(
  RedemptionsDatasyncEvents.VAULT_CREATED,
  z.string(),
  VaultCreatedEventDetailSchema,
);
export type VaultCreatedEvent = z.infer<typeof VaultCreatedEventSchema>;
export type VaultCreatedEventDetail = z.infer<typeof VaultCreatedEventDetailSchema>;

export class VaultCreatedController extends EventBridgeController<VaultCreatedEvent> {
  static readonly inject = [Logger.key, VaultService.key] as const;

  constructor(
    logger: ILogger,
    protected vaultService: IVaultService,
  ) {
    super(logger);
  }

  protected parseRequest(request: UnknownEventBridgeEvent): Result<VaultCreatedEvent, Error> {
    // TODO: Remove this log message after migration
    this.logger.info({
      message: 'Parsing request',
      context: {
        request,
      },
    });
    return this.zodParseRequest(request, VaultCreatedEventSchema);
  }

  protected async handle(event: VaultCreatedEvent): Promise<void> {
    await this.vaultService.createVault(event);
  }
}
