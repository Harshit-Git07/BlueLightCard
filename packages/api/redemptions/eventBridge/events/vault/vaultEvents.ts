export enum VaultEvents {
  VAULT_CREATED = 'vault.created',
  VAULT_UPDATED = 'vault.updated',
}
