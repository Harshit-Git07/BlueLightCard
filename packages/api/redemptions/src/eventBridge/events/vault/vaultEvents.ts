import { z } from 'zod';

import { NUMERIC_ID_SCHEMA } from '../../../schemas/common';
import { PLATFORM_SCHEMA } from '../../../schemas/domain';
import { EventSchema, eventSchema } from '../../schemas/event';

export enum VaultEvents {
  VAULT_CREATED = 'vault.created',
  VAULT_UPDATED = 'vault.updated',
}

export const VaultEventSchema = {
  [VaultEvents.VAULT_CREATED]: eventSchema(
    VaultEvents.VAULT_CREATED,
    z.object({
      adminEmail: z.string().email().nullable().optional(),
      alertBelow: z.number(),
      brand: z.string(),
      companyId: NUMERIC_ID_SCHEMA,
      companyName: z.string(),
      eeCampaignId: NUMERIC_ID_SCHEMA.nullable().optional(),
      link: z.string().url().nullable().optional(),
      linkId: NUMERIC_ID_SCHEMA.nullable().optional(),
      managerId: NUMERIC_ID_SCHEMA.nullable().optional(),
      maxPerUser: NUMERIC_ID_SCHEMA,
      offerId: NUMERIC_ID_SCHEMA,
      platform: PLATFORM_SCHEMA,
      showQR: z.boolean(),
      terms: z.string(),
      ucCampaignId: NUMERIC_ID_SCHEMA.nullable().optional(),
      vaultStatus: z.boolean(),
    }),
  ),
  [VaultEvents.VAULT_UPDATED]: eventSchema(VaultEvents.VAULT_UPDATED, z.object({})),
} satisfies Record<VaultEvents, EventSchema>;

export const VaultCreatedEventSchema = VaultEventSchema[VaultEvents.VAULT_CREATED];
export type VaultCreatedEvent = z.infer<typeof VaultCreatedEventSchema>;
export type VaultCreatedEventDetail = VaultCreatedEvent['detail'];

export const VaultUpdatedEventSchema = VaultEventSchema[VaultEvents.VAULT_UPDATED];
export type VaultUpdatedEvent = z.infer<typeof VaultUpdatedEventSchema>;
export type VaultUpdatedEventDetail = VaultUpdatedEvent['detail'];
