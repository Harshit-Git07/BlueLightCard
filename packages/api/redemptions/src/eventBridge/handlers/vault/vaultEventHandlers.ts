// eslint-disable-next-line @typescript-eslint/no-empty-function -- TODO: Implementation
export const vaultCreatedHandler = (): void => {};
// eslint-disable-next-line @typescript-eslint/no-empty-function -- TODO: Implementation
export const vaultUpdatedHandler = (): void => {};
