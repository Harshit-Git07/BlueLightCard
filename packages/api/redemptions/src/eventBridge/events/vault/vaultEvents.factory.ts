import { faker } from '@faker-js/faker';
import { Factory } from 'fishery';

import { VaultCreatedEvent, VaultCreatedEventDetail } from './vaultEvents';

export const vaultCreatedEventDetailFactory = Factory.define<VaultCreatedEventDetail>(() => ({
  adminEmail: faker.helpers.maybe(() => faker.internet.email()),
  alertBelow: faker.number.int(500),
  brand: faker.company.name(),
  companyId: faker.number.int({
    min: 1,
    max: 1_000_000,
  }),
  companyName: faker.company.name(),
  directCode: faker.helpers.maybe(() => faker.string.alphanumeric(10)),
  eeCampaignId: faker.helpers.maybe(() =>
    faker.number.int({
      min: 1,
      max: 1_000_000,
    }),
  ),
  link: faker.helpers.maybe(() => faker.internet.url()),
  linkId: faker.helpers.maybe(() =>
    faker.number.int({
      min: 1,
      max: 1_000_000,
    }),
  ),
  managerId: faker.helpers.maybe(() =>
    faker.number.int({
      min: 1,
      max: 1_000_000,
    }),
  ),
  maxPerUser: faker.number.int(500),
  offerId: faker.number.int({
    min: 1,
    max: 1_000_000,
  }),
  platform: faker.helpers.arrayElement(['BLC_UK', 'BLC_AU', 'DDS_UK']),
  showQR: faker.datatype.boolean(),
  terms: faker.lorem.paragraph(),
  ucCampaignId: faker.helpers.maybe(() =>
    faker.number.int({
      min: 1,
      max: 1_000_000,
    }),
  ),
  vaultId: faker.string.uuid(),
  vaultStatus: faker.datatype.boolean(),
}));
export const vaultCreatedEventFactory = Factory.define<VaultCreatedEvent>(() => ({
  account: faker.string.numeric(12),
  detail: vaultCreatedEventDetailFactory.build(),
  'detail-type': 'vault.created',
  id: faker.string.uuid(),
  region: faker.helpers.arrayElement(['eu-west-1', 'ap-west-2', 'us-east-1']),
  resources: ['vault.created'],
  source: 'vault.created',
  time: faker.date.recent().toISOString(),
  version: '0',
}));
