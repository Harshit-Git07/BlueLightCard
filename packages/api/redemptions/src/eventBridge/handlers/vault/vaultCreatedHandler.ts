import { sql } from 'drizzle-orm';

import { ILogger } from '@blc-mono/core/src/utils/logger/logger';

import { DatabaseConnection } from '../../../database/connection';
import { createVaultId, redemptionsTable, vaultsTable } from '../../../database/schema';
import { VaultCreatedEventDetail, VaultCreatedEventSchema } from '../../events/vault/vaultEvents';

type NewVault = typeof vaultsTable.$inferInsert;

function parseIntegration(event: VaultCreatedEventDetail): Pick<NewVault, 'integration' | 'integrationId'> {
  switch (true) {
    case Boolean(event.eeCampaignId):
      return {
        integration: 'eagleeye',
        integrationId: event.eeCampaignId ?? null,
      };
    case Boolean(event.ucCampaignId):
      return {
        integration: 'uniqodo',
        integrationId: event.ucCampaignId ?? null,
      };
    default:
      return {
        integration: null,
        integrationId: null,
      };
  }
}

export const vaultCreatedHandler = async ({ db }: DatabaseConnection, logger: ILogger, event: unknown) => {
  const vaultCreatedEvent = VaultCreatedEventSchema.parse(event);
  const eventDetail = vaultCreatedEvent.detail;

  const newVault: Omit<NewVault, 'redemptionId'> = {
    alertBelow: eventDetail.alertBelow,
    status: eventDetail.vaultStatus ? 'active' : 'in-active',
    maxPerUser: eventDetail.maxPerUser,
    showQR: eventDetail.showQR,
    terms: eventDetail.terms,
    email: eventDetail.adminEmail || null,
    ...parseIntegration(eventDetail),
  };

  const query = sql`
    INSERT INTO ${vaultsTable} (
      "${sql.raw(vaultsTable.id.name)}",
      "${sql.raw(vaultsTable.redemptionId.name)}",
      "${sql.raw(vaultsTable.alertBelow.name)}",
      "${sql.raw(vaultsTable.status.name)}",
      "${sql.raw(vaultsTable.maxPerUser.name)}",
      "${sql.raw(vaultsTable.showQR.name)}",
      "${sql.raw(vaultsTable.terms.name)}",
      "${sql.raw(vaultsTable.email.name)}",
      "${sql.raw(vaultsTable.integration.name)}",
      "${sql.raw(vaultsTable.integrationId.name)}"
    )
    SELECT
      ${createVaultId()},
      ${redemptionsTable.id},
      ${newVault.alertBelow},
      ${newVault.status},
      ${newVault.maxPerUser},
      ${newVault.showQR},
      ${newVault.terms},
      ${newVault.email},
      ${newVault.integration},
      ${newVault.integrationId}
    FROM ${redemptionsTable}
    WHERE (
      ${redemptionsTable.companyId} = ${eventDetail.companyId}
      AND ${redemptionsTable.offerId} = ${eventDetail.offerId}
      AND ${redemptionsTable.platform} = ${eventDetail.platform}
    )
    RETURNING (
      ${vaultsTable.id}
    )
  `;
  const results = await db.execute(query);

  if (results.length === 0) {
    logger.error({
      message: 'No redemption found for vault',
      context: {
        offerId: eventDetail.offerId,
        companyId: eventDetail.companyId,
        platform: eventDetail.platform,
      },
    });
  }

  if (results.length > 1) {
    logger.error({
      message: 'Multiple redemptions found for vault',
      context: {
        offerId: eventDetail.offerId,
        companyId: eventDetail.companyId,
        platform: eventDetail.platform,
      },
    });
  }
};
