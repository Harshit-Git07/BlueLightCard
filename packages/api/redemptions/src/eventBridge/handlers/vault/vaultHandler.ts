import { LambdaLogger } from '@blc-mono/core/utils/logger/lambdaLogger';
import { ILogger } from '@blc-mono/core/utils/logger/logger';

import { DatabaseConnection, DatabaseConnectionType } from '../../../database/connection';
import { withConnection } from '../../../database/connectionHelpers';
import { VaultEvents } from '../../events';
import { IEventBridge } from '../../types/event';

import { vaultCreatedHandler } from './vaultCreatedHandler';
import { vaultUpdatedHandler } from './vaultUpdatedHandler';

type VaultCreatedHandler = typeof vaultCreatedHandler;
type VaultUpdatedHandler = typeof vaultUpdatedHandler;

export function createHandler(
  logger: ILogger,
  vaultCreatedHandler: VaultCreatedHandler,
  vaultUpdatedHandler: VaultUpdatedHandler,
) {
  return async function vaultHandler(connection: DatabaseConnection, event: IEventBridge): Promise<void> {
    const { source } = event;

    switch (source) {
      case VaultEvents.VAULT_CREATED:
        vaultCreatedHandler(connection, logger, event);
        break;
      case VaultEvents.VAULT_UPDATED:
        vaultUpdatedHandler();
        break;
      default:
        logger.error({ message: `Invalid Event  ${source}` });
        throw new Error('Invalid event source');
    }
  };
}

const logger = new LambdaLogger({ serviceName: `redemptions-offer-event-handler` });

export const handler = withConnection(
  DatabaseConnectionType.READ_WRITE,
  createHandler(logger, vaultCreatedHandler, vaultUpdatedHandler),
);
