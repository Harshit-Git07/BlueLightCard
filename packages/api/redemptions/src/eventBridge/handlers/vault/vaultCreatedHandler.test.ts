import { faker } from '@faker-js/faker';
import { describe, it } from '@jest/globals';
import { QueryWithTypings } from 'drizzle-orm';
import { PgDialect } from 'drizzle-orm/pg-core';

import { as } from '@blc-mono/core/utils/testing';

import { createTestLogger } from '../../../helpers/test/logger';
import { VaultCreatedEvent } from '../../events/vault/vaultEvents';
import { vaultCreatedEventFactory } from '../../events/vault/vaultEvents.factory';

import { vaultCreatedHandler } from './vaultCreatedHandler';

describe('vaultCreatedHandler', () => {
  const logger = createTestLogger();

  describe('should map event data correctly', () => {
    function test(
      name: string,
      event: VaultCreatedEvent,
      expected?: (event: VaultCreatedEvent, query: QueryWithTypings) => void,
    ) {
      it(name, async () => {
        // Arrange
        const mockDatabaseConnection = {
          db: {
            execute: jest.fn().mockReturnValueOnce([
              {
                id: 'mock-vault-id',
              },
            ]),
          },
        };

        // Act
        await vaultCreatedHandler(as(mockDatabaseConnection), logger, event);

        // Assert
        expect(mockDatabaseConnection.db.execute).toHaveBeenCalledTimes(1);
        const pgDialect = new PgDialect();
        const query = pgDialect.sqlToQuery(mockDatabaseConnection.db.execute.mock.calls[0][0]);
        expected?.(event, query);
        expect(query.sql).toMatchSnapshot();
      });
    }
    test(
      'for Eagle Eye vaults',
      vaultCreatedEventFactory.build({
        detail: {
          eeCampaignId: faker.number.int(500),
          ucCampaignId: undefined,
        },
      }),
      (event, query) => {
        expect(query.params).toContain('eagleeye');
        expect(query.params).toContain(event.detail.eeCampaignId);
      },
    );

    test(
      'for Uniqodo vaults',
      vaultCreatedEventFactory.build({
        detail: {
          eeCampaignId: undefined,
          ucCampaignId: faker.number.int(500),
        },
      }),
      (event, query) => {
        expect(query.params).toContain('uniqodo');
        expect(query.params).toContain(event.detail.ucCampaignId);
      },
    );

    test(
      'for active vaults',
      vaultCreatedEventFactory.build({
        detail: {
          vaultStatus: true,
        },
      }),
      (_, query) => {
        expect(query.params).toContain('active');
      },
    );

    test(
      'for inactive vaults',
      vaultCreatedEventFactory.build({
        detail: {
          vaultStatus: false,
        },
      }),
      (_, query) => {
        expect(query.params).toContain('in-active');
      },
    );

    test(
      'for admin email',
      vaultCreatedEventFactory.build({
        detail: {
          adminEmail: faker.internet.email(),
        },
      }),
      (event, query) => {
        expect(query.params).toContain(event.detail.adminEmail);
      },
    );

    test(
      'for missing admin email',
      vaultCreatedEventFactory.build({
        detail: {
          adminEmail: undefined,
        },
      }),
    );

    test('for passthrough fields', vaultCreatedEventFactory.build(), (event, query) => {
      expect(query.params).toContain(event.detail.alertBelow);
      expect(query.params).toContain(event.detail.maxPerUser);
      expect(query.params).toContain(event.detail.showQR);
      expect(query.params).toContain(event.detail.terms);
    });
  });

  describe('throws an error if the event is invalid', () => {
    function test(name: string, event: unknown) {
      it(name, async () => {
        // Arrange
        const mockDatabaseConnection = {
          db: {
            execute: jest.fn().mockReturnValue([
              {
                id: 'mock-vault-id',
              },
            ]),
          },
        };

        // Act & Assert
        expect(
          async () => await vaultCreatedHandler(as(mockDatabaseConnection), logger, event),
        ).rejects.toThrowErrorMatchingSnapshot();
        expect(mockDatabaseConnection.db.execute).not.toHaveBeenCalled();
      });
    }

    test('because the event is null', null);
    test('because the event is undefined', undefined);
    test('because the event is an empty object', {});
    test('because the event is missing detail', vaultCreatedEventFactory.build({ detail: undefined }));
    test('because the event is missing time', vaultCreatedEventFactory.build({ time: undefined }));
    test('because the event detail is an empty object', {
      ...vaultCreatedEventFactory.build(),
      detail: {},
    });
    test(
      'because the `adminEmail` field is invalid',
      vaultCreatedEventFactory.build({
        detail: {
          adminEmail: 'not-an-email',
        },
      }),
    );
    test(
      'because the `alertBelow` field is invalid',
      vaultCreatedEventFactory.build({
        detail: {
          alertBelow: '123' as unknown as number,
        },
      }),
    );
    test(
      'because the `brand` field is invalid',
      vaultCreatedEventFactory.build({
        detail: {
          brand: 123 as unknown as string,
        },
      }),
    );
    test(
      'because the `companyId` field is invalid',
      vaultCreatedEventFactory.build({
        detail: {
          companyId: '123' as unknown as number,
        },
      }),
    );
    test(
      'because the `companyName` field is invalid',
      vaultCreatedEventFactory.build({
        detail: {
          companyName: false as unknown as string,
        },
      }),
    );
    test(
      'because the `eeCampaignId` field is invalid',
      vaultCreatedEventFactory.build({
        detail: {
          eeCampaignId: '123' as unknown as number,
        },
      }),
    );
    test(
      'because the `link` field is invalid',
      vaultCreatedEventFactory.build({
        detail: {
          link: 123 as unknown as string,
        },
      }),
    );
    test(
      'because the `linkId` field is invalid',
      vaultCreatedEventFactory.build({
        detail: {
          linkId: '123' as unknown as number,
        },
      }),
    );
    test(
      'because the `managerId` field is invalid',
      vaultCreatedEventFactory.build({
        detail: {
          managerId: '123' as unknown as number,
        },
      }),
    );
    test(
      'because the `maxPerUser` field is invalid',
      vaultCreatedEventFactory.build({
        detail: {
          maxPerUser: '123' as unknown as number,
        },
      }),
    );
    test(
      'because the `offerId` field is invalid',
      vaultCreatedEventFactory.build({
        detail: {
          offerId: '123' as unknown as number,
        },
      }),
    );
    test(
      'because the `platform` field is invalid',
      vaultCreatedEventFactory.build({
        detail: {
          platform: 'BLC' as 'BLC_UK',
        },
      }),
    );
    test(
      'because the `showQR` field is invalid',
      vaultCreatedEventFactory.build({
        detail: {
          showQR: 'true' as unknown as boolean,
        },
      }),
    );
    test(
      'because the `terms` field is invalid',
      vaultCreatedEventFactory.build({
        detail: {
          terms: false as unknown as string,
        },
      }),
    );
    test(
      'because the `ucCampaignId` field is invalid',
      vaultCreatedEventFactory.build({
        detail: {
          ucCampaignId: '123' as unknown as number,
        },
      }),
    );
    test(
      'because the `vaultStatus` field is invalid',
      vaultCreatedEventFactory.build({
        detail: {
          vaultStatus: 'true' as unknown as boolean,
        },
      }),
    );
  });
});
