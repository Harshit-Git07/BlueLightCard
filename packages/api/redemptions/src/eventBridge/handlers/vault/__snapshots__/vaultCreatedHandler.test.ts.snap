// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vaultCreatedHandler should map event data correctly for Eagle Eye vaults 1`] = `
"
    INSERT INTO "vaults" (
      "id",
      "redemptionId",
      "alertBelow",
      "status",
      "maxPerUser",
      "showQR",
      "terms",
      "email",
      "integration",
      "integrationId"
    )
    SELECT
      $1,
      "redemptions"."id",
      $2,
      $3,
      $4,
      $5,
      $6,
      $7,
      $8,
      $9
    FROM "redemptions"
    WHERE (
      "redemptions"."companyId" = $10
      AND "redemptions"."offerId" = $11
      AND "redemptions"."platform" = $12
    )
    RETURNING (
      "vaults"."id"
    )
  "
`;

exports[`vaultCreatedHandler should map event data correctly for Uniqodo vaults 1`] = `
"
    INSERT INTO "vaults" (
      "id",
      "redemptionId",
      "alertBelow",
      "status",
      "maxPerUser",
      "showQR",
      "terms",
      "email",
      "integration",
      "integrationId"
    )
    SELECT
      $1,
      "redemptions"."id",
      $2,
      $3,
      $4,
      $5,
      $6,
      $7,
      $8,
      $9
    FROM "redemptions"
    WHERE (
      "redemptions"."companyId" = $10
      AND "redemptions"."offerId" = $11
      AND "redemptions"."platform" = $12
    )
    RETURNING (
      "vaults"."id"
    )
  "
`;

exports[`vaultCreatedHandler should map event data correctly for active vaults 1`] = `
"
    INSERT INTO "vaults" (
      "id",
      "redemptionId",
      "alertBelow",
      "status",
      "maxPerUser",
      "showQR",
      "terms",
      "email",
      "integration",
      "integrationId"
    )
    SELECT
      $1,
      "redemptions"."id",
      $2,
      $3,
      $4,
      $5,
      $6,
      $7,
      $8,
      $9
    FROM "redemptions"
    WHERE (
      "redemptions"."companyId" = $10
      AND "redemptions"."offerId" = $11
      AND "redemptions"."platform" = $12
    )
    RETURNING (
      "vaults"."id"
    )
  "
`;

exports[`vaultCreatedHandler should map event data correctly for admin email 1`] = `
"
    INSERT INTO "vaults" (
      "id",
      "redemptionId",
      "alertBelow",
      "status",
      "maxPerUser",
      "showQR",
      "terms",
      "email",
      "integration",
      "integrationId"
    )
    SELECT
      $1,
      "redemptions"."id",
      $2,
      $3,
      $4,
      $5,
      $6,
      $7,
      $8,
      $9
    FROM "redemptions"
    WHERE (
      "redemptions"."companyId" = $10
      AND "redemptions"."offerId" = $11
      AND "redemptions"."platform" = $12
    )
    RETURNING (
      "vaults"."id"
    )
  "
`;

exports[`vaultCreatedHandler should map event data correctly for inactive vaults 1`] = `
"
    INSERT INTO "vaults" (
      "id",
      "redemptionId",
      "alertBelow",
      "status",
      "maxPerUser",
      "showQR",
      "terms",
      "email",
      "integration",
      "integrationId"
    )
    SELECT
      $1,
      "redemptions"."id",
      $2,
      $3,
      $4,
      $5,
      $6,
      $7,
      $8,
      $9
    FROM "redemptions"
    WHERE (
      "redemptions"."companyId" = $10
      AND "redemptions"."offerId" = $11
      AND "redemptions"."platform" = $12
    )
    RETURNING (
      "vaults"."id"
    )
  "
`;

exports[`vaultCreatedHandler should map event data correctly for missing admin email 1`] = `
"
    INSERT INTO "vaults" (
      "id",
      "redemptionId",
      "alertBelow",
      "status",
      "maxPerUser",
      "showQR",
      "terms",
      "email",
      "integration",
      "integrationId"
    )
    SELECT
      $1,
      "redemptions"."id",
      $2,
      $3,
      $4,
      $5,
      $6,
      $7,
      $8,
      $9
    FROM "redemptions"
    WHERE (
      "redemptions"."companyId" = $10
      AND "redemptions"."offerId" = $11
      AND "redemptions"."platform" = $12
    )
    RETURNING (
      "vaults"."id"
    )
  "
`;

exports[`vaultCreatedHandler should map event data correctly for passthrough fields 1`] = `
"
    INSERT INTO "vaults" (
      "id",
      "redemptionId",
      "alertBelow",
      "status",
      "maxPerUser",
      "showQR",
      "terms",
      "email",
      "integration",
      "integrationId"
    )
    SELECT
      $1,
      "redemptions"."id",
      $2,
      $3,
      $4,
      $5,
      $6,
      $7,
      $8,
      $9
    FROM "redemptions"
    WHERE (
      "redemptions"."companyId" = $10
      AND "redemptions"."offerId" = $11
      AND "redemptions"."platform" = $12
    )
    RETURNING (
      "vaults"."id"
    )
  "
`;

exports[`vaultCreatedHandler throws an error if the event is invalid because the \`adminEmail\` field is invalid 1`] = `
"[
  {
    "validation": "email",
    "code": "invalid_string",
    "message": "Invalid email",
    "path": [
      "detail",
      "adminEmail"
    ]
  }
]"
`;

exports[`vaultCreatedHandler throws an error if the event is invalid because the \`alertBelow\` field is invalid 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [
      "detail",
      "alertBelow"
    ],
    "message": "Expected number, received string"
  }
]"
`;

exports[`vaultCreatedHandler throws an error if the event is invalid because the \`brand\` field is invalid 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "detail",
      "brand"
    ],
    "message": "Expected string, received number"
  }
]"
`;

exports[`vaultCreatedHandler throws an error if the event is invalid because the \`companyId\` field is invalid 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [
      "detail",
      "companyId"
    ],
    "message": "Expected number, received string"
  }
]"
`;

exports[`vaultCreatedHandler throws an error if the event is invalid because the \`companyName\` field is invalid 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "boolean",
    "path": [
      "detail",
      "companyName"
    ],
    "message": "Expected string, received boolean"
  }
]"
`;

exports[`vaultCreatedHandler throws an error if the event is invalid because the \`eeCampaignId\` field is invalid 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [
      "detail",
      "eeCampaignId"
    ],
    "message": "Expected number, received string"
  }
]"
`;

exports[`vaultCreatedHandler throws an error if the event is invalid because the \`link\` field is invalid 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "number",
    "path": [
      "detail",
      "link"
    ],
    "message": "Expected string, received number"
  }
]"
`;

exports[`vaultCreatedHandler throws an error if the event is invalid because the \`linkId\` field is invalid 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [
      "detail",
      "linkId"
    ],
    "message": "Expected number, received string"
  }
]"
`;

exports[`vaultCreatedHandler throws an error if the event is invalid because the \`managerId\` field is invalid 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [
      "detail",
      "managerId"
    ],
    "message": "Expected number, received string"
  }
]"
`;

exports[`vaultCreatedHandler throws an error if the event is invalid because the \`maxPerUser\` field is invalid 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [
      "detail",
      "maxPerUser"
    ],
    "message": "Expected number, received string"
  }
]"
`;

exports[`vaultCreatedHandler throws an error if the event is invalid because the \`offerId\` field is invalid 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [
      "detail",
      "offerId"
    ],
    "message": "Expected number, received string"
  }
]"
`;

exports[`vaultCreatedHandler throws an error if the event is invalid because the \`platform\` field is invalid 1`] = `
"[
  {
    "received": "BLC",
    "code": "invalid_enum_value",
    "options": [
      "BLC_UK",
      "BLC_AU",
      "DDS_UK"
    ],
    "path": [
      "detail",
      "platform"
    ],
    "message": "Invalid enum value. Expected 'BLC_UK' | 'BLC_AU' | 'DDS_UK', received 'BLC'"
  }
]"
`;

exports[`vaultCreatedHandler throws an error if the event is invalid because the \`showQR\` field is invalid 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "boolean",
    "received": "string",
    "path": [
      "detail",
      "showQR"
    ],
    "message": "Expected boolean, received string"
  }
]"
`;

exports[`vaultCreatedHandler throws an error if the event is invalid because the \`terms\` field is invalid 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "boolean",
    "path": [
      "detail",
      "terms"
    ],
    "message": "Expected string, received boolean"
  }
]"
`;

exports[`vaultCreatedHandler throws an error if the event is invalid because the \`ucCampaignId\` field is invalid 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "string",
    "path": [
      "detail",
      "ucCampaignId"
    ],
    "message": "Expected number, received string"
  }
]"
`;

exports[`vaultCreatedHandler throws an error if the event is invalid because the \`vaultStatus\` field is invalid 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "boolean",
    "received": "string",
    "path": [
      "detail",
      "vaultStatus"
    ],
    "message": "Expected boolean, received string"
  }
]"
`;

exports[`vaultCreatedHandler throws an error if the event is invalid because the event detail is an empty object 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      "detail",
      "alertBelow"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "detail",
      "brand"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      "detail",
      "companyId"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "detail",
      "companyName"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      "detail",
      "maxPerUser"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "number",
    "received": "undefined",
    "path": [
      "detail",
      "offerId"
    ],
    "message": "Required"
  },
  {
    "expected": "'BLC_UK' | 'BLC_AU' | 'DDS_UK'",
    "received": "undefined",
    "code": "invalid_type",
    "path": [
      "detail",
      "platform"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "boolean",
    "received": "undefined",
    "path": [
      "detail",
      "showQR"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "detail",
      "terms"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "boolean",
    "received": "undefined",
    "path": [
      "detail",
      "vaultStatus"
    ],
    "message": "Required"
  }
]"
`;

exports[`vaultCreatedHandler throws an error if the event is invalid because the event is an empty object 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "account"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "undefined",
    "path": [
      "detail"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "detail-type"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "id"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "region"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "array",
    "received": "undefined",
    "path": [
      "resources"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "source"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "time"
    ],
    "message": "Required"
  },
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "version"
    ],
    "message": "Required"
  }
]"
`;

exports[`vaultCreatedHandler throws an error if the event is invalid because the event is missing detail 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "undefined",
    "path": [
      "detail"
    ],
    "message": "Required"
  }
]"
`;

exports[`vaultCreatedHandler throws an error if the event is invalid because the event is missing time 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "string",
    "received": "undefined",
    "path": [
      "time"
    ],
    "message": "Required"
  }
]"
`;

exports[`vaultCreatedHandler throws an error if the event is invalid because the event is null 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "null",
    "path": [],
    "message": "Expected object, received null"
  }
]"
`;

exports[`vaultCreatedHandler throws an error if the event is invalid because the event is undefined 1`] = `
"[
  {
    "code": "invalid_type",
    "expected": "object",
    "received": "undefined",
    "path": [],
    "message": "Required"
  }
]"
`;
